exposestrategy/auto.go

```diff
@@ -0,0 +1,129 @@
+package exposestrategy
+
+import (
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+const (
+	ingress      = "route"
+	loadBalancer = "loadbalancer"
+	nodePort     = "nodeport"
+	route        = "route"
+	domainExt    = ".xip.io"
+)
+
+func NewAutoStrategy(exposer, domain string, client *client.Client, restClientConfig *restclient.Config, encoder runtime.Encoder) (ExposeStrategy, error) {
+
+	exposer, err := getAutoDefaultExposeRule(client)
+	if err != nil {
+		errors.Wrap(err, "failed to automatically get exposer rule.  consider setting 'exposer' type in config.yml")
+	} else {
+		glog.Infof("Using exposer strategy: %s", exposer)
+	}
+
+	// only try to get domain if we need wildcard dns and one wasn't given to us
+	if len(domain) == 0 && (strings.EqualFold(ingress, exposer) || strings.EqualFold(route, exposer)) {
+
+		domain, err = getAutoDefaultDomain(client)
+		if err != nil {
+			errors.Wrap(err, "failed to get a domain")
+		} else {
+			glog.Infof("Using domain: %s", domain)
+		}
+
+	}
+
+	return New(exposer, domain, client, restClientConfig, encoder)
+}
+
+func getAutoDefaultExposeRule(c *client.Client) (string, error) {
+
+	nodes, err := c.Nodes().List(api.ListOptions{})
+	if err != nil {
+		errors.Wrap(err, "failed to find any nodes")
+	}
+	if len(nodes.Items) == 1 {
+		node := nodes.Items[0]
+		if node.Name == "minishift" || node.Name == "minikube" {
+			return nodePort, nil
+		}
+	}
+
+	t, err := typeOfMaster(c)
+	if err != nil {
+		return "", err
+	}
+	if t == kubernetes {
+		return ingress, nil
+	} else if t == openShift {
+		return route, nil
+	}
+	return "", errors.New("no known automatic ways to get an external ip to use with xip")
+}
+
+func getAutoDefaultDomain(c *client.Client) (string, error) {
+	nodes, err := c.Nodes().List(api.ListOptions{})
+	if err != nil {
+		errors.Wrap(err, "failed to find any nodes")
+	}
+
+	// if we're mini* then there's only one node, any router / ingress controller deployed has to be on this one
+	if len(nodes.Items) == 1 {
+		node := nodes.Items[0]
+		if node.Name == "minishift" || node.Name == "minikube" {
+			ip, err := getExternalIP(node)
+			if err != nil {
+
+				return "", err
+			}
+			return ip + domainExt, nil
```

Comment:
domainExt is a const = ".xip.io", if this is needs to be overridden then user can add a domain value to the config.  readme has been updated to explain the default of xip